Lazily initialize native crypto libraries

JasonFengJ9 opened the pull request:
Delay native crypto library loading to avoid early initialization which might cause CRIU cross architectures restore error; 
Skip native library loading before taking a checkpoint;
Reload the providers from scratch;
Create a singleton NativeCrypto instance.

fix https://github.com/OpenLiberty/open-liberty/issues/23360

Signed-off-by: Jason Feng <fengj@ca.ibm.com>

tajila commented:
@keithc-ca please review these changes

keithc-ca commented on closed/src/java.base/share/classes/jdk/crypto/jniprovider/NativeCrypto.java:
```diff
@@ -129,33 +160,24 @@ public static final boolean isAlgorithmEnabled(String property, String name, boo
-                    System.err.println("Warning: Native crypto library load failed." +
-                            " Using Java crypto implementation.");
+                    System.err.println("Warning: " + name + " native requirements not satisfied. " +
+                        explanation + " Using Java crypto implementation.");
```
Please indent continued lines 8 spaces (like it was before).

JasonFengJ9 replied:
Updated

keithc-ca commented on closed/src/java.base/share/classes/jdk/crypto/jniprovider/NativeCrypto.java:
```diff
@@ -178,6 +200,7 @@ public static NativeCrypto getNativeCrypto() {
 
         if ((callerClassLoader == null) || (callerClassLoader == VM.getVMLangAccess().getExtClassLoader())) {
+            initCryptoLibraries();
             return new NativeCrypto();
```
This should be a singleton given the new usage pattern.

JasonFengJ9 replied:
Changed it to a singleton instance.

JasonFengJ9 commented on src/java.base/share/classes/com/sun/crypto/provider/CipherCore.java:
```diff
@@ -183,7 +183,7 @@ void setMode(String mode) throws NoSuchAlgorithmException {
              * the NativeCipherBlockChaining class.
              */
-            if (useNativeCBC && blockSize == 16) {
+            if (useNativeCBC && NativeCrypto.isLoaded() && blockSize == 16) {
```
Updated

JasonFengJ9 commented on src/java.base/share/classes/com/sun/crypto/provider/PKCS12PBECipherCore.java:
```diff
@@ -119,6 +119,13 @@ static byte[] derive(char[] chars, byte[] salt, int ic, int n, int type,
             }
             if (hashSupported) {
+                if (nativeCrypto == null) {
+                    synchronized (PKCS12PBECipherCore.class) {
```
The synchronization is not needed after creating a singleton `NativeCrypto` instance.

JasonFengJ9 commented on src/java.base/share/classes/sun/security/rsa/RSACore.java:
```diff
@@ -117,7 +117,7 @@ public static byte[] convert(byte[] b, int ofs, int len) {
             throws BadPaddingException {
-        if (useNativeRsa && key instanceof sun.security.rsa.RSAPublicKeyImpl) {
+        if (useNativeRsa && NativeCrypto.isLoaded() && key instanceof sun.security.rsa.RSAPublicKeyImpl) {
             byte[] ret = NativeRSACore.rsa(msg, (sun.security.rsa.RSAPublicKeyImpl) key);
```
Updated

keithc-ca replied:
I didn't mean to just add an import (I don't think that's helpful for one use).
I suggested pattern matching.

JasonFengJ9 replied:
Ah, I misinterpreted it and missed the pattern variable.
Fixed.

JasonFengJ9 commented on src/java.base/share/classes/sun/security/rsa/RSAPrivateCrtKeyImpl.java:
```diff
@@ -356,7 +359,9 @@ protected void returnNativePtr(long ptr) {
-            nativeCrypto.destroyRSAKey(itr);
+            if (nativeCrypto != null) {
+                nativeCrypto.destroyRSAKey(itr);
+            }
```
If `nativeCrypto == null`, `returnNativePtr()` shouldn't be invoked hence no entries in `keyQ`. In that case it is sufficient for implicit null check via current `nativeCrypto.destroyRSAKey(itr)`.

JasonFengJ9 commented on closed/src/java.base/share/classes/jdk/crypto/jniprovider/NativeCrypto.java:
```diff
@@ -53,33 +53,64 @@ public class NativeCrypto {
+    private static void initCryptoLibraries() {
+        // double checking ossl_ver to ensure only one OpenSSL library loading
+        if (ossl_ver == -9) {
+            synchronized(NativeCrypto.class) {
```
Added

JasonFengJ9 commented:
Added a change to `SunEC` which delays native libraries loading, and removed `jdk.nativeEC=false` workaround introduced earlier for OpenLiberty before https://github.com/OpenLiberty/open-liberty/issues/23360 was reported.

@keithc-ca this is ready for another look.

keithc-ca commented on src/jdk.crypto.ec/share/classes/sun/security/ec/SunEC.java:
```diff
@@ -182,7 +182,7 @@ public Object newInstance(Object ctrParamObj)
                 } else  if (type.equals("KeyAgreement")) {
                     if (algo.equals("ECDH")) {
-                        if (useNativeEC) {
+                        if (useNativeEC && NativeCrypto.isLoaded()) {
```
A similar change should be made on line 351.

JasonFengJ9 replied:
L351 can proceed w/ `useNativeEC` enabled, `NativeCrypto.isLoaded()` is needed only when a `NativeCripto` usage is imminent.
Invoking `NativeCrypto.isLoaded()` at `SunEC` instantiation stage caused a similar issue like https://github.com/OpenLiberty/open-liberty/issues/23360 before.

keithc-ca commented:
> Pattern matching was done.

Perhaps you have a different understanding of "pattern matching", but it isn't being used as suggested; see https://github.com/ibmruntimes/openj9-openjdk-jdk17/pull/157#discussion_r1038385668.

keithc-ca commented:
> Added a change to `SunEC` which delays native libraries loading, this removes `jdk.nativeEC=false` workaround introduced earlier for OpenLiberty before [OpenLiberty/open-liberty#23360](https://github.com/OpenLiberty/open-liberty/issues/23360) was reported.
> 
> @keithc-ca this is ready for another look.

I think the test removed from `SunEC` was necessary. Otherwise, `NativeECDHKeyAgreement` can call `NativeCrypto.ECNativeGF2m()` without loading the native code (which I expect will result in an `UnsatisfiedLinkError`).

keithc-ca commented on src/jdk.crypto.ec/share/classes/sun/security/ec/SunEC.java:
```diff
@@ -182,7 +186,13 @@ public Object newInstance(Object ctrParamObj)
                 } else  if (type.equals("KeyAgreement")) {
                     if (algo.equals("ECDH")) {
-                        if (useNativeEC) {
+                        boolean isNativeCryptoLoadAllowed = true;
+/*[IF CRIU_SUPPORT]*/
+                        if (InternalCRIUSupport.isCheckpointAllowed()) {
+                            isNativeCryptoLoadAllowed = false;
+                        }
+/*[ENDIF] CRIU_SUPPORT*/
+                        if (useNativeEC && isNativeCryptoLoadAllowed && NativeCrypto.isLoaded()) {
```
Suggest we don't need the extra local variable:
```java
                        if (useNativeEC
/*[IF CRIU_SUPPORT]*/
                            && !InternalCRIUSupport.isCheckpointAllowed()
/*[ENDIF] CRIU_SUPPORT */
                            && NativeCrypto.isLoaded()
                        ) {
```

JasonFengJ9 replied:
Updated

JasonFengJ9 commented on closed/src/java.base/share/classes/jdk/crypto/jniprovider/NativeCrypto.java:
```diff
@@ -52,34 +52,67 @@ public class NativeCrypto {
-                    }
-                    // signal load failure
-                    ossl_ver = -1;
+    private static int ossl_ver = -1;
```
Removed

JasonFengJ9 commented on src/jdk.crypto.ec/share/classes/sun/security/ec/SunEC.java:
```diff
@@ -56,6 +56,10 @@
 
+/*[IF CRIU_SUPPORT]*/
+import openj9.internal.criu.InternalCRIUSupport;
+/*[ENDIF] CRIU_SUPPORT*/
```
Yeah, fixed.

JasonFengJ9 commented on src/java.base/share/classes/com/sun/crypto/provider/SunJCE.java:
```diff
@@ -292,7 +292,7 @@ void putEntries() {
         attrs.put("SupportedKeyFormats", "RAW");
 
-        if (useNativeGaloisCounterMode) {
+        if (useNativeGaloisCounterMode && NativeCrypto.isLoaded()) {
```
This `SunJCE` provider wasn't instantiated before `CRIUSECProvider` replacing it and other providers, so it didn't cause the issue * https://github.com/OpenLiberty/open-liberty/issues/23360.
Similarly `NativeCrypto` APIs such as `isLoaded()`, `getVersion()` and `getNativeCrypto()` have usages that _could_ trigger native library loading but are avoided by `CRIUSECProvider`, I can address those usages in a separate PR instead.

JasonFengJ9 replied:
Added the check `!InternalCRIUSupport.isCheckpointAllowed()` before `NativeCrypto.getVersion() >= 1` & `NativeCrypto.isLoaded()` at `SunJCE`/`SunEntries`, `getNativeCrypto()` usages are either not in the constructing path or in dedicated native crypto classes which are only instantiated after such usages are allowed.
Removed `NativeCrypto.isAlgorithmEnabled(String, String, boolean, String)`.

@keithc-ca this is ready for another look.

keithc-ca commented on closed/src/java.base/share/classes/jdk/crypto/jniprovider/NativeCrypto.java:
```diff
@@ -110,47 +125,24 @@ public static final boolean isAlgorithmEnabled(String property, String name) {
-                    System.err.println("Warning: Native crypto library load failed." +
-                            " Using Java crypto implementation.");
-                }
+            if (traceEnabled) {
+                System.err.println(name + " native crypto implementation enabled.");
             }
         } else {
             if (traceEnabled) {
```
Suggest reversing the nesting here:
```java
        if (traceEnabled) {
            if (useNativeAlgorithm) {
...
```

JasonFengJ9 replied:
Updated

keithc-ca commented on src/java.base/share/classes/com/sun/crypto/provider/SunJCE.java:
```diff
@@ -337,7 +345,13 @@ void putEntries() {
+            && !InternalCRIUSupport.isCheckpointAllowed()
+/*[ENDIF] CRIU_SUPPORT */
+            && (NativeCrypto.getVersion() >= 1)
+            && NativeCrypto.isLoaded()
```
There's no need to call `NativeCrypto.isLoaded()`; it's implied by `NativeCrypto.getVersion() >= 1`.

JasonFengJ9 replied:
Removed

keithc-ca commented on src/java.base/share/classes/com/sun/crypto/provider/SunJCE.java:
```diff
@@ -292,7 +295,12 @@ void putEntries() {
-        if (useNativeGaloisCounterMode) {
+        if (useNativeGaloisCounterMode
+/*[IF CRIU_SUPPORT]*/
+            && !InternalCRIUSupport.isCheckpointAllowed()
```
This is problematic: I think it disables the native implementation even after a restore because the Java implementation will be registered before the checkpoint (both GCM and ChaCha20).

JasonFengJ9 replied:
There is no problem, `CRIUSECProvider` replaces other providers before taking checkpoint, and put them back after restore, all provider instances have to be recreated, the previously registered service will have been removed and re-registered. See `CRIUConfigurator.setCRIUSecMode()/setCRIURestoreMode()`.

keithc-ca replied:
If this provider is not used before a checkpoint, then why does it need to ask `isCheckpointAllowed()`?

keithc-ca commented on src/java.base/share/classes/sun/security/provider/SunEntries.java:
```diff
@@ -278,7 +282,12 @@ public final class SunEntries {
-        if (useNativeDigest) {
+        if (useNativeDigest
+/*[IF CRIU_SUPPORT]*/
+            && !InternalCRIUSupport.isCheckpointAllowed()
```
Like in `SunJCE`, I think this disables the native implementation even after a restore.

JasonFengJ9 replied:
Same as

JasonFengJ9 commented:
@keithc-ca ping

keithc-ca commented:
Please create a pull request with the desired changes for the head stream, after which back-ports will be considered.

JasonFengJ9 commented:
@keithc-ca this has been updated with https://github.com/ibmruntimes/openj9-openjdk-jdk/pull/539, please review.

keithc-ca commented:
Jenkins test sanity xmac jdk17
